src/ImageSharp/Formats/ImageDecoder.cs

```diff
@@ -0,0 +1,254 @@
+    }
+
+    internal static Task<T> WithSeekableMemoryStreamAsync<T>(
+        DecoderOptions options,
+        Stream stream,
+        Func<Stream, CancellationToken, T> action,
+        CancellationToken cancellationToken)
+    {
+        Guard.NotNull(options, nameof(options));
+        Guard.NotNull(stream, nameof(stream));
+
+        if (!stream.CanRead)
+        {
+            throw new NotSupportedException("Cannot read from the stream.");
+        }
+
+        Task<T> PeformActionAndResetPosition(Stream s, long position, CancellationToken ct)
+        {
+            try
+            {
+                T result = action(s, ct);
+
+                // Issue #2259. Our buffered reads may have left the stream in an incorrect non-zero position.
+                // Reset the position of the seekable stream if we did not read to the end to allow additional reads.
+                // We check here that the input stream is seekable because it is not guaranteed to be so since
+                // we always copy input streams of unknown type.
+                if (stream.CanSeek && stream.Position != s.Position && s.Position != s.Length)
+                {
+                    stream.Position = position + s.Position;
+                }
+
+                return Task.FromResult(result);
+            }
+            catch (OperationCanceledException)
+            {
+                return Task.FromCanceled<T>(cancellationToken);
+            }
+            catch (Exception ex)
+            {
+                return Task.FromException<T>(ex);
+            }
+        }
+
+        // NOTE: We are explicitly not executing the action against the stream here as we do in WithSeekableStream() because that
+        // would incur synchronous IO reads which must be avoided in this asynchronous method. Instead, we will *always* run the
+        // code below to copy the stream to an in-memory buffer before invoking the action.
+        if (stream is MemoryStream ms)
+        {
+            return PeformActionAndResetPosition(ms, ms.Position, cancellationToken);
+        }
+
+        if (stream is ChunkedMemoryStream cms)
+        {
+            return PeformActionAndResetPosition(cms, cms.Position, cancellationToken);
+        }
+
+        return CopyToMemoryStreamAndActionAsync(options, stream, PeformActionAndResetPosition, cancellationToken);
+    }
+
+    private static async Task<T> CopyToMemoryStreamAndActionAsync<T>(
+        DecoderOptions options,
+        Stream stream,
+        Func<Stream, long, CancellationToken, Task<T>> action,
+        CancellationToken cancellationToken)
+    {
+        long position = stream.CanSeek ? stream.Position : 0;
+        Configuration configuration = options.Configuration;
+        using ChunkedMemoryStream memoryStream = new(configuration.MemoryAllocator);
+        await stream.CopyToAsync(memoryStream, configuration.StreamProcessingBufferSize, cancellationToken).ConfigureAwait(false);
```

Comment:
@antonfirsov  What I don't understand about the test that is timing out is that the entire stream is copied here in the failing condition. This means we shouldn't even reach the actual decode operation before the position-based cancellation kicks in and an exception is thrown.

Reply:
The position based cancellation will kick in during the copy, since the copy is utilizing reads/writes since 3a7c4f442d22ba7370bed4b49b772f01e096f274. When `PausedMemoryStream` is used `WithSeekableMemoryStreamAsync` will threat it as a `MemoryStream` and avoid wrapping it.

We'll need to press the rerun a few more times to prove it, but I think 85dc0f3 fixed the test issues. I introduced the timeout to handle a scenario where I noticed (or imagined?) that the paused stream was hanging without actually trying to understand why did it hang. Anyways, good news is that I don't see this hang happening anymore.

Instead, we were seeing the tests timing out because a decoder run can actually take longer than 30 seconds on a heavily overloaded 1-core VM.

Reply:
Our code shouldn't, (_at least I couldn't find a reason it should following refactoring_) be hanging since the wrapping since the sync wrapping handles exceptions and always returns a Task. 

Fingers crossed this has it! I'll trigger the run again now.

Reply:
That's 2 successful runs in a row. I'll trigger a 3rd now.

Reply:
I also did it 3 times, should be good now hopefully.